### ISSUE TITLE
Improve Add::__hash__() and Mul::__hash__()

### ISSUE THREAD

**Issue by isuruf** [symengine/symengine#683]
This came up in gitter a while ago, but was not looked into.

Current implementation of `Add::__hash__()` orders the elements first and then hashes. A better algorithm would be to add the hashes together. Benchmarks would have to be run to see if there is any difference.

**Comment by mafiya69** [comment]
What about this : 
```patch
 std::size_t Add::__hash__() const
 {
     std::size_t seed = ADD;
-    hash_combine<Basic>(seed, *coef_);
-    map_basic_num ordered(dict_.begin(), dict_.end());
-    for (const auto &p: ordered) {
-        hash_combine<Basic>(seed, *(p.first));
-        hash_combine<Basic>(seed, *(p.second));
+    seed += coef_->__hash__();
+    for(const auto &p : dict_) {
+      seed += p.first->__hash__();
+      seed += p.second->__hash__();
     }
     return seed;
 }
```

**Comment by isuruf** [comment]
That should work.
Can you check if there's any performance improvement?

**Comment by mafiya69** [comment]
Previously the running time of `/benchmarks/add1` is mostly `48xx`, and after the above patch it is mostly `47xx`.

**Comment by isuruf** [comment]
Closed via #718
